source/val/validate_builtins.cpp

```diff
@@ -2909,139 +2891,180 @@ spv_result_t BuiltInsValidator::ValidateVertexIndexAtReference(
   return SPV_SUCCESS;
 }
 
-spv_result_t BuiltInsValidator::ValidateMeshEXTBuiltinInterfaceRules(
+typedef struct {
+  uint32_t array_type;
+  uint32_t array_size;
+  uint32_t block_array_size;
+  uint32_t perprim_deco;
+} MeshBuiltinVUIDs;
+
+// clang-format off
+std::unordered_map<spv::BuiltIn, MeshBuiltinVUIDs> MeshBuiltinVUIDMap = {{
+    {spv::BuiltIn::CullPrimitiveEXT,        {7036, 10589, 10590, 7038}},
+    {spv::BuiltIn::PrimitiveId,             {10595, 10596, 10597, 7040}},
+    {spv::BuiltIn::Layer,                   {10592, 10593, 10594, 7039}},
+    {spv::BuiltIn::ViewportIndex,           {10601, 10602, 10603, 7060}},
+    {spv::BuiltIn::PrimitiveShadingRateKHR, {10598, 10599, 10600, 7059}},
+}};
+// clang-format on
+
+spv_result_t BuiltInsValidator::ValidateMeshBuiltinInterfaceRules(
     const Decoration& decoration, const Instruction& inst, spv::Op scalar_type,
-    uint32_t array_type_vuid, uint32_t array_size_vuid,
-    uint32_t block_array_size_vuid, uint32_t interface_vuid,
-    uint32_t perprim_deco_vuid) {
-  bool is_block = false;
-  const spv::BuiltIn builtin = decoration.builtin();
-  // Validate the type of the builtin and if it's part of a block or an array
-  if (spv_result_t error = ValidateBlockTypeOrArrayedType(
-          decoration, inst, is_block, scalar_type,
-          [this, &inst, &decoration,
-           &array_type_vuid](const std::string& message) -> spv_result_t {
-            return _.diag(SPV_ERROR_INVALID_DATA, &inst)
-                   << _.VkErrorID(array_type_vuid) << "According to the "
-                   << spvLogStringForEnv(_.context()->target_env)
-                   << " spec BuiltIn "
-                   << _.grammar().lookupOperandName(
-                          SPV_OPERAND_TYPE_BUILT_IN,
-                          (uint32_t)decoration.builtin())
-                   << " variable needs to be a either a boolean or an "
-                      "array of booleans."
-                   << message;
-          })) {
-    return error;
-  }
+    const Instruction& referenced_from_inst) {
+  if (function_id_) {
+    if (execution_models_.count(spv::ExecutionModel::MeshEXT)) {
+      bool is_block = false;
+      const spv::BuiltIn builtin = decoration.builtin();
+      const MeshBuiltinVUIDs& vuids = MeshBuiltinVUIDMap[builtin];
+      if (spv_result_t error = ValidateBlockTypeOrArrayedType(
+              decoration, inst, is_block, scalar_type,
+              [this, &inst, &builtin, &scalar_type,
+               &vuids](const std::string& message) -> spv_result_t {
+                return _.diag(SPV_ERROR_INVALID_DATA, &inst)
```

Comment:
print out the `scalar_type` now (getting https://gitlab.khronos.org/vulkan/vulkan/-/merge_requests/7554 to confirm it should be how we have it in currently in the code)